README
Probix graph server using Common Lisp


ChangeLog
2008-12-07 Sergey Sinkovskiy <glorybox.away@gmail.com>

	* graph-server.lisp: removed commented out condition
	* vecto-chart.lisp: fixed scaling,
	fixed start point

2008-12-07 Sergey Sinkovskiy <glorybox.away@gmail.com>

	* added scaling
	* minor fixes



1.lisp
(defpackage #:test
  (:use #:cl #:vecto))

(in-package #:test)

(defun fac (n)
  (if (= n 0)
      1
      (* n (fac (decf n)))))



(defun feedlike-icon (file size-x size-y)
  (with-canvas (:width size-x :height size-y)
    (set-rgb-fill 1.0 0.65 0.3)
    (rounded-rectangle 0 0 size-x size-y 10 10)
    (fill-path)
    (set-rgb-fill 1.0 1.0 1.0)
    (centered-circle-path 20 20 10)
    (fill-path)
    (flet ((quarter-circle (x y radius)
             (move-to (+ x radius) y)
             (arc x y radius 0 (/ pi 2))))
      (set-rgb-stroke 1.0 1.0 1.0)
      (set-line-width (/ size-x 7))
      (quarter-circle 20 20 (/ size-x 2.5))
      (stroke)
      (quarter-circle 20 20 (/ size-x 1.5))
      (stroke))
    (rounded-rectangle 5 5 (- size-x 10) (- size-y 10) 7 7)
    (set-gradient-fill 50 90
                       1.0 1.0 1.0 0.7
                       50 20
                       1.0 1.0 1.0 0.0)
    (set-line-width 2)
    (set-rgba-stroke 1.0 1.0 1.0 0.1)
    (fill-and-stroke)
    (save-png file)))




test-page.lisp
(require 'cl-who)
(require 'hunchentoot)
(require 'vecto)
(require 'cl-json)

(defpackage :test-page
  (:use :cl :cl-who :hunchentoot :vecto :json :flexi-streams))

(in-package :test-page)

(defun demo-handler ()
  (with-html-output-to-string
      (*standard-output* nil :prologue t :indent nil)
    (:html
     (:head
      (:title "Probix graph server"))
     (:body
      (:p "Pass the JSON here:")
      (:form :method "post"
             (:textarea :name "json" :cols 60 :rows 15 (format t "~A" (post-parameter "json")))
             (:br)
             (:input :type :submit))
      (:p
       (format t "POST: ~A" (post-parameters)))
      (:p
       (let ((decoded-json nil))
         (ignore-errors
           (setf decoded-json (decode-json-from-string
                               (post-parameter "json"))))
         (format t "DECODED JSON: ~A"
                 (if (not decoded-json)
                     "Error in decoding JSON"
                     decoded-json))))
      (:p
       (:img :src "/demo/graph.png"))))))

(defun build-graph ()
  (setf (content-type) "image/png")
  (with-canvas (:width 100 :height 100)
    (set-rgb-fill 1.0 0.65 0.3)
    (rounded-rectangle 0 0 100 100 10 10)
    (fill-path)
    (set-rgb-fill 1.0 1.0 1.0)
    (centered-circle-path 20 20 10)
    (fill-path)
    (flet ((quarter-circle (x y radius)
             (move-to (+ x radius) y)
             (arc x y radius 0 (/ pi 2))))
      (set-rgb-stroke 1.0 1.0 1.0)
      (set-line-width 15)
      (quarter-circle 20 20 30)
      (stroke)
      (quarter-circle 20 20 60)
      (stroke))
    (rounded-rectangle 5 5 90 90 7 7)
    (set-gradient-fill 50 90
                       1.0 1.0 1.0 0.7
                       50 20
                       1.0 1.0 1.0 0.0)
    (set-line-width 2)
    (set-rgba-stroke 1.0 1.0 1.0 0.1)
    (fill-and-stroke)
    (flexi-streams:with-output-to-sequence (png-stream)
      (save-png-stream png-stream)
      png-stream)))

;    (let ((png-stream (make-instance 'sb-gray:fundamental-binary-input-stream)))
;      (save-png-stream png-stream)
;      png-stream)))

(push (create-prefix-dispatcher "/demo" 'demo-handler) *dispatch-table*)
(push (create-prefix-dispatcher "/demo/graph.png" 'build-graph) *dispatch-table*)



vecto-chart.lisp
(cl:defpackage :vecto-chart
  (:use :cl :vecto :flexi-streams)
  (:export #:render-png-stream #:add-data #:with-line-chart #:draw-axis))

(in-package :vecto-chart)

(defclass chart ()
  ((width :accessor chart-width :initarg :width :initform 400)
   (height :accessor chart-height :initarg :height :initform 300)
   (data :accessor chart-data :initarg :data :initform nil)
   (offset :accessor chart-offset :initarg :offset :initform 0))))

(defclass line-chart (chart)
  ((have-x-y :accessor chart-have-x-y :initform nil)
   (min-x :accessor chart-data-min-x :initform 0)
   (min-y :accessor chart-data-min-y :initform 0)
   (max-x :accessor chart-data-max-x :initform 0)
   (max-y :accessor chart-data-max-y :initform 0)
   (ratio-x :accessor chart-ratio-x :initform 1)
   (ratio-y :accessor chart-ratio-y :initform 1)))

(defclass pie-chart (chart)
  ())


(defmacro with-pie-chart ((&key width height) &body body)
 `(let ((*current-chart* (make-instance 'pie-chart
				      :width ,width
				      :height ,height)))
	(with-canvas (:width ,width :height ,height)
	  ,@body)))

(defvar *current-chart* nil)

(defmacro with-line-chart ((&key width height) &body body)
  `(let ((*current-chart* (make-instance 'line-chart
										 :width ,width
										 :height ,height
										 :offset (* ,height 0.05))))
	 (with-canvas (:width ,width :height ,height)
	   ,@body)))



(defparameter +series-colors+
  (list '(0.6 0.5 0.9)
        '(0.3 0.6 0.7)
        '(0.5 0.3 0.8)
        '(0.9 0.4 0.3)
        '(0.4 0.8 0.2)))

(defun draw-axis ()
  (draw-axis* *current-chart*))

(defmethod draw-axis* ((chart line-chart))
  (with-slots (width height offset) chart
    (set-rgb-fill 0.95 0.95 0.95)
    (rectangle 0 0 width height)
    (fill-path)
    (set-line-cap :square)
	;; (set-line-width 0.35)
    (move-to offset offset)
    (line-to offset (- height offset))
    (move-to offset offset)
    (line-to (- width offset) offset)
    (stroke)))


(define-condition wrong-argument-structure (condition)
  ())

;;; helper that uses lexical variable as a chart instance
(defun add-data (data)
  (add-data* *current-chart* data))

;;; CHART method to add more data to chart
;;; It validates data for consistency and
;;; pushes it to data slot
(defmethod add-data* ((chart chart) data)
  ;; add data validation here

  ;; 1) checking if data is empty
  (if (eq data nil)
	  (cond ((every #'numberp data)
			 (setf (chart-have-x-y chart) nil))
			((every #'consp data)
			 (setf (chart-have-x-y chart) T))))

  ;; 2) checking what we have list of numbers or list of conses
  ;; and checking if this is consistent with data that we have
  (if (or 
	   (and (every #'numberp data) (chart-have-x-y chart))
	   (and (every #'consp data) (not (chart-have-x-y chart)))
	   (not (or (not (every #'numberp data)) (not (every #'consp data)))))
	  (error 'wrong-argument-structure))
  
  (push data (chart-data chart)))


;;; Macro to avoid lot of boilerplate code
;;; that generates CHART class method
(defmacro calculate-min-max* (&body body)
  `(defmethod calculate-min-max ((chart line-chart))
	 (unless (chart-have-x-y chart)
	   ,@(loop for fn in '(min max)
			  collect
			  `(setf (,(intern (format nil "CHART-DATA-~A-Y" fn)) chart)
					 (apply #',fn (apply #'append (chart-data chart)))))
	   (setf (chart-data-min-x chart) 1)
	   (setf (chart-data-max-x chart)
			 (apply #'max
					(mapcar #'length (chart-data chart)))))
	 (when (chart-have-x-y chart)
	   ,@(loop for fn in '(min max)
		   collect 
			 `(setf (,(intern (format nil "CHART-DATA-~A-X" fn)) chart)
					(apply #',fn
						   (apply #'car (apply #'append (chart-data chart)))))
		   collect
			 `(setf (,(intern (format nil "CHART-DATA-~A-Y" fn)) chart)
					 (apply #',fn
							(apply #'cdr
								   (apply #'append (chart-data chart)))))))))

;; generating method
(calculate-min-max*)

;;; VECTO method that calculates ratios for x and y axis
(defmethod calculate-ratios ((chart line-chart))
  (with-slots (width height offset ratio-x ratio-y min-x min-y max-x max-y) chart
	(setf ratio-x
		  (/ (- width
				(* 2 offset) ; from left and right
				10) ; offset from the end of X axis
			 (- max-x min-x)))
	(setf ratio-y
		  (/ (- height
				(* 2 offset) ; from top and bottom
				10) ; from the end of Y axis
			 (- max-y min-y )))))


;;; Method to convert data values to pixel values for X axis
(defmethod dx->x ((chart line-chart) x)
  (+ (chart-offset chart) 
	 (* (chart-ratio-x chart)
		(- x (chart-data-min-x chart)))))

;;; Method to convert data values to pixel values for Y axis
(defmethod dy->y ((chart line-chart) y)
  (+ (chart-offset chart)
	 (* (chart-ratio-y chart)
		(- y (chart-data-min-y chart)))))


;;; short helper
(defun render-png-stream ()
  (render-png-stream* *current-chart*))

;;; Main drawing method
;;; returns png data stream
(defmethod render-png-stream* ((chart line-chart))
  (let ((data (chart-data chart))
		(have-x-y (chart-have-x-y chart)))
	;; calculating min-max-values for all data
	(calculate-min-max chart)
	;; calculate ratios for given chart
	(calculate-ratios chart)

	;; iterating through data and drawing line
	(loop for dataset in data do
		 (if have-x-y
			 (move-to
			  (1+ (dx->x chart (caar dataset)))
			  (1+ (dy->y chart (cadr dataset))))
			 (move-to
			  (1+ (dx->x chart 1))
			  (1+ (dy->y chart (car dataset)))))
		 (pop dataset)
		 (apply #'set-rgb-stroke
				(nth (random (length +series-colors+)) +series-colors+))
		 (set-line-width 1)
		 (set-line-join :round)
		 (set-line-cap :round)
		 (let ((current-x 2))
		   (if have-x-y
			   (loop for item in dataset do
					(line-to
					 (dx->x chart (car item))
					 (dy->y chart (cdr item))))
			   (loop for item in dataset do
					(line-to
					 (dx->x chart current-x)
					 (dy->y chart item))
					(incf current-x))))
		 (stroke))
	(flexi-streams:with-output-to-sequence (png-stream)
	  (save-png-stream png-stream)
	  png-stream)))



graph-server.lisp
(defpackage :graph-server
  (:use :cl :cl-who :hunchentoot :vecto :json :flexi-streams :vecto-chart)
  (:export start-server stop-server))

(in-package :graph-server)

(defun json-form-handler ()
    (with-html-output-to-string
      (*standard-output* nil :prologue t :indent nil)
    (:html
     (:head
      (:title "Probix graph server"))
     (:body
      (:p "Pass the JSON here:")
      (:form :method "post" :action "/image"
             (:textarea :name "json" :cols 60 :rows 15 (format nil "~A" (post-parameter "json")) "{
 \"size\": {
   \"x\": 500,
    \"y\": 300
 },
 \"data\": [10,40,50,60,40,50,60,70,40,30]}")
             (:br)
             (:select :name "output-format"
                      (:option :value "png" "PNG") (:option :value "debug" "DEBUG"))
             (:br)
             (:input :type :submit))))))

(defun parse-json (json)
  (json-bind (size data) json
    (json-bind (x y) size
      (if (not (and x y data (listp data)))
          (error 'json-parse-error :message
                 (format nil "Size: ~A, x: ~A, y: ~A, data: ~A" size x y data))
          (list :data data :size-x x :size-y y)))))

(defun json-post-handler ()
  (let (decoded-json)
    (handler-case
        (progn 
          (setf decoded-json (parse-json (post-parameter "json")))
          (cond ((string= (post-parameter "output-format") "png")
                 (image-output-png decoded-json))
                ((string= (post-parameter "output-format") "debug")
                 (image-output-debug decoded-json))))
      (json-parse-error (perror)
        (output-error-json perror)))))

(defun output-error-json (err)
    (setf (content-type) "text/plain")
    (with-html-output-to-string
        (*standard-output* nil :prologue nil :indent nil)
      (format t "Error: ~A" err)))

(defun image-output-png (json)
  (setf (content-type) "image/png")
  (let ((data (getf json :data))
        (size-x (getf json :size-x))
        (size-y (getf json :size-y)))
    (with-line-chart (:width size-x :height size-y)
	  (draw-axis)
	  (add-data data)
	  (add-data '(10 50 60 40 30 70 60 70 80 60))
	  (add-data '(120 40 50 70 60 40 40 50 60))
	  (render-png-stream))))

  
(defun image-output-debug (json)
  (with-html-output-to-string (*standard-output* nil :prologue nil :indent nil)
    (format t "DECODED JSON: ~A" json)))

(defun json-data-handler ()
  "[[10, 20], [20, 30], [30, 40], [40,20]]")

(push (create-prefix-dispatcher "/json-form" 'json-form-handler) *dispatch-table*)
(push (create-prefix-dispatcher "/image" 'json-post-handler) *dispatch-table*)
(push (create-prefix-dispatcher "/json-data" 'json-data-handler) *dispatch-table*)

